parso/python/errors.py

```diff
@@ -1008,3 +1033,82 @@ def is_issue(self, for_stmt):
         expr_list = for_stmt.children[1]
         if expr_list.type != 'expr_list':  # Already handled.
             self._check_assignment(expr_list)
+
+
+@ErrorFinder.register_rule(type='namedexpr_test')
+class _NamedExprRule(_CheckAssignmentRule):
+    # namedexpr_test: test [':=' test]
+
+    def is_issue(self, namedexpr_test):
+        first = namedexpr_test.children[0]
+        if first.type == 'lambdef':
+            # (lambda: x := 1)
+            self.add_issue(namedexpr_test, message='cannot use named assignment with lambda')
+        elif first.type == 'atom_expr':
+            for child in first.children:
+                if child.type != 'trailer':
+                    continue
+                first_child = child.children[0]
+                if first_child.type == 'operator':
+                    if first_child.value == '[':
+                        # (a[i] := x)
+                        self.add_issue(namedexpr_test, message='cannot use named assignment with subscript')
+                    elif first_child.value == '.':
+                        # (a.b := c)
+                        self.add_issue(namedexpr_test, message='cannot use named assignment with attribute')
+        else:
+            self._check_assignment(first, is_namedexpr=True)
+
+
+@ErrorFinder.register_rule(type='testlist_comp')
+@ErrorFinder.register_rule(type='dictorsetmaker')
+class _ComprehensionRule(SyntaxRule):
```

Comment:
Can we replace this with parent accesses in `_NamedExprRule`? There's also `search_ancestor` that might help with it. Not 100% sure if it works, but I feel like it's nicer than to scan for named exprs below.

Reply:
Okay, it's all done.